rfc/rfc-0014-source-code-upload.md

```diff
@@ -0,0 +1,66 @@
+# RFC-0011: Source Code Upload
+
+**Authors:** Emily Casey
+
+**Status:**
+
+**Pull Request URL:** [#1374](https://github.com/projectriff/riff/pull/1374)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+The way we currently handle local source is less than ideal in a few ways:
+1. Images built from local source will not get autmatically rebased or rebuilt based on stack or buildpack updates,
+ and are therefore susceptible to more security vulnerability
+1. Because build from source uses `pack`, it requires users to have a docker daemon
+1. It requires users to wait for large builder images to pull
+1. Results produced by pack and kpack should be identical but difference introduces risk
+
+## Solution
+
+#### Registry Type SourceResolver
+When a user specifies the `--local-path <dir>` flag during `riff function create` or `riff application create` the source
+code in the specified directory will be packaged as a OCI image containing a single layer and uploaded to a docker registry.
+`riff` with create a `kpack` `image` with a `registry` type `sourceresolver` https://github.com/pivotal/kpack/blob/master/docs/image.md#source-configuration.
+`kpack` can then build the image as it does in the git source type case.
+
+#### Source Image Tag
+This will require a repository to which riff can upload the source image. By default `riff` can upload the source image to a predictable 
+derivation of the target image tag, by appending a suffix to the tag. For example if the generated image will be published to `docker.io/my/app:latest`,
+the source image will be uploaded to `docker.io/my/app:latest-source`
+
+The following flags will allow for configurability:
+
+* `--source-image repository`
+
+#### Source Image Push Auth
+When pushing the source image `riff` will use the default docker keychain and will not try to use the registry secrets configured in `riff`
+
+#### Source Image Pull Auth
+Credentials for the source image must be present in the riff service account so that kpack can pull the source code
```

Comment:
This sounds like an additional manual step for someone using non-public registry like GCR or can we add something to the `riff credentials` command?

Reply:
I believe the existing `riff credentials` commands should suffice. If I understand the `kpack` `build-init` keychain logic correctly, it should include all of the registry credentials from the service account.

The only thing users would need to do is apply another set of registry creds, if the registry they are using to store their source code is different from that to which they publish. If they are using the same regsistry for both images, no additional steps should be required.

Reply:
the `riff credential` commands are already able to handle secrets from multiple registries

Reply:
@scothis I think we are saying the same thing